api/src/auth/auth_utils.py

```diff
@@ -1,8 +1,76 @@
-from typing import Any
+import functools
+import logging
+from typing import Any, Callable, ParamSpec
+
+import flask
+from apiflask.exceptions import HTTPError
+
+from src.api import response
+from src.auth.login_gov_jwt_auth import get_final_redirect_uri
+
+logger = logging.getLogger(__name__)
+
+P = ParamSpec("P")
+INTERNAL_ERROR = "internal error"
 
 
 def get_app_security_scheme() -> dict[str, Any]:
     return {
         "ApiKeyAuth": {"type": "apiKey", "in": "header", "name": "X-Auth"},
         "ApiJwtAuth": {"type": "apiKey", "in": "header", "name": "X-SGG-Token"},
     }
+
+
+def with_login_redirect_error_handler() -> Callable[..., Callable[P, flask.Response]]:
+    """Wrapper function to handle catching errors and redirecting
+
+    Because several of our login functions don't have standard 2xx returns
+    and instead redirect the user, we also redirect in the case of errors
+    so that they stay on the frontend, but we pass errors along.
+
+    Usage::
+
+        @with_login_redirect_error_handler()
+        def foo(...):
+            logger.info("hello")
+
+            if condition:
+                raise_flask_error(...) # this will get caught and a redirect will occur
+
+            return ...
+
+    """
+
+    def decorator(f: Callable[P, flask.Response]) -> Callable[P, flask.Response]:
+        @functools.wraps(f)
+        def wrapper(*args: P.args, **kwargs: P.kwargs) -> flask.Response:
+            try:
+                return f(*args, **kwargs)
+            except HTTPError as e:
+                # HTTPError is what raise_flask_error raises
+                # and should encompass our "expected" errors
+                # that aren't a concern, as long as it isn't a 5xx
+                message = e.message
+                logger.info("Login flow failed")
```

Comment:
Do we want to include the message here so it's clearer what logs here correspond to what users are experiencing?

Reply:
I think I realized that about two PRs after this one and adjusted some of the logging here, I'll backport it after I merge the upstream changes